Cython/Compiler/ExprNodes.py

```diff
@@ -14796,7 +14798,16 @@ def check_const(self):
         return self.arg.check_const()
 
     def calculate_result_code(self):
-        return "(%s != 0)" % self.arg.result()
+        arg = self.arg.result()
+        if self.arg.type.is_string:
+            # Test for 0-length string with "ptr[0] != '\0'" instead of just "ptr != 0".
+            # If we'd coerce the pointer to a Python 'bytes' / 'str', we'd do the equivalent
+            # but much less efficiently (strlen -> string object -> len>0).
+            # Unicode strings might fail to decode (so it's not entirely equivalent), but the code
```

Comment:
Unless I'm missing something, I don't think it'd ever try to decode beyond the `'\0'` byte anyway

Reply:
It's a behavioural difference for non-empty strings though. If the string is incorrectly (non-UTF8) encoded then decoding it to determine the size would fail, whereas if we only look at the first byte and do not do any decoding then we'd decide that it's not empty but without producing the failure. That's what I meant.